lib/Dialect/RTG/IR/RTGTypes.cpp

```diff
@@ -19,6 +19,77 @@ using namespace rtg;
 #define GET_TYPEDEF_CLASSES
 #include "circt/Dialect/RTG/IR/RTGTypes.cpp.inc"
 
+//===----------------------------------------------------------------------===//
+// DictType
+//===----------------------------------------------------------------------===/
+
+LogicalResult DictType::verify(function_ref<InFlightDiagnostic()> emitError,
+                               ArrayRef<StringAttr> entryNames,
+                               ArrayRef<Type> entryTypes) {
+  if (entryNames.size() != entryTypes.size())
+    return emitError() << "must have the same number of names and types";
+
+  llvm::SmallDenseSet<StringAttr> entryNameSet;
+  LogicalResult result = success();
+  entryNameSet.reserve(entryNames.size());
+  for (auto name : entryNames) {
+    if (!entryNameSet.insert(name).second) {
+      result = failure();
+      emitError() << "duplicate entry name '" << name.getValue()
+                  << "' in rtg.dict type";
+    }
+  }
+  return result;
+}
+
+Type DictType::parse(AsmParser &p) {
+  llvm::StringSet<> nameSet;
```

Comment:
Since we're interning the field names anyways, I think it would make more sense to use some kind of StringAttrs instead.